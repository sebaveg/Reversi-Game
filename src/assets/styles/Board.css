
.Board {
  background-color: #8756ca;
  border: 3px solid white;
  box-shadow: 0 0 15px -5px #fff;
  min-width: 280px;
  margin-top: 1em;
}

.wrapper-btn-undo-redo{
  display: flex;
  justify-content: space-between;
  margin-top: .5em;
}

.wrapper-btn-undo-redo button{
  background-color: #8756ca;
  border: 1px solid #8756ca;
  box-shadow: 0 0 5px -2px #fff;
  color: #fff;
  cursor: pointer;
  font-size: 1.5rem;
  font-weight: bold;
  padding: 5px 20px;
}

.wrapper-btn-undo-redo button:disabled {
  visibility: hidden;
}