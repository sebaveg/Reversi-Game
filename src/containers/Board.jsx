import React from 'react';
import { ActionCreators as UndoActionCreators } from 'redux-undo';
import { connect } from 'react-redux';
import {
  setBoard,
  setAllowedCells,
  addDisksPlayers,
  changeTurn,
} from '../actions';

import BoardLayout from '../components/Board';

const directions = [
  [0, 1], // right
  [0, -1], // left
  [-1, 0], // up
  [1, 0], // down
  [1, 1], // down right
  [1, -1], // down left
  [-1, 1], // up right
  [-1, -1], // up left
];

class Board extends React.Component {
  constructor(props) {
    super(props);
    this.rows = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'];
    this.cols = ['1', '2', '3', '4', '5', '6', '7', '8'];
  }

  componentDidMount() {
    this.initialBoard();
  }

  componentDidUpdate(prevProps) {
    const passTurn = this.props.currentPlayer !== prevProps.currentPlayer
    || this.props.posDiskWhite.length !== prevProps.posDiskWhite.length
    || this.props.posDiskBlack.length !== prevProps.posDiskBlack.length;
    if (passTurn) {
      const newBoard = this.allowedCellsAndCountDisks(this.props.board);
      this.props.setBoard(newBoard); // dispatch action
    }
  }

  diskOponent = () => (this.props.currentPlayer === 'white' ? 'black' : 'white');

  inBoard = (x, y) => ((x >= 0) && (x < 8) && (y >= 0) && (y < 8));

  createBoard() {
    const board = this.rows.slice();
    for (let i = 0; i < 8; i += 1) {
      board[i] = this.cols.map((row) => ({
        id: row + board[i],
        disk: null, // white or black
        allowedCell: false, // says if enable or disable click for put disk
      }));
    }
    return board;
  }

  async initialBoard() {
    // Set initial disks: two white and two black in center
    const board = this.createBoard();
    this.props.posDiskWhite.forEach((pos) => {
      board[pos[0]][pos[1]].disk = 'white'; // White
    });
    this.props.posDiskBlack.forEach((pos) => {
      board[pos[0]][pos[1]].disk = 'black'; // Black
    });
    const newBoard = this.allowedCellsAndCountDisks(board);
    await this.props.setBoard(newBoard); // dispatch action
  }

  // says what cells are enable for clicks in current turn
  allowedCellsAndCountDisks(board) {
    const b = board.slice();
    let arrayCanReverse;
    let cantAllow = 0; let disksWhite = 0; let disksBlack = 0;
    for (let x = 0; x < 8; x++) {
      for (let y = 0; y < 8; y++) {
        // array with positions enable for put disks)
        arrayCanReverse = this.canPutDisk(board, x, y);
        b[x][y].allowedCell = arrayCanReverse;
        if (arrayCanReverse.length > 0) cantAllow += 1;
        // count disks
        if (b[x][y].disk === 'white') disksWhite += 1;
        if (b[x][y].disk === 'black') disksBlack += 1;
      }
    }
    this.props.setAllowedCells(cantAllow); // dispatch how many cells enable
    this.props.addDisksPlayers({
      white: disksWhite, black: disksBlack,
    }); // dispatch how many disks for player
    return b;
  }

  // array with positions enable for put disks
  canPutDisk(board, x, y) {
    const b = board.slice();
    if (b[x][y].disk) return [];
    const canReverse = [];
    let X;
    let Y;
    let cantDisks;
    let cells;
    directions.forEach((direction) => {
      cantDisks = 0;
      cells = [];
      X = x;
      Y = y;
      do {
        cells.push({ X, Y });
        X += direction[0];
        Y += direction[1];
        cantDisks += 1;
      } while (this.inBoard(X, Y) && b[X][Y].disk === this.diskOponent());

      if (cantDisks > 1 && this.inBoard(X, Y) && b[X][Y].disk === this.props.currentPlayer) {
        canReverse.push(cells);
        // cells.forEach((cell) => {
        //   board[cell[0]][cell[1]].disk = this.props.currentPlayer;
        // });
      }
    });
    return [].concat(...canReverse);
  }

  async handleUndo() {
    await this.props.onUndo();
    // await this.props.changeTurn();
  }

  async handleRedo() {
    await this.props.onRedo();
    // await this.props.changeTurn();
  }

  render() {
    return (
      <>
        <BoardLayout board={this.props.board} />
        <button type="button" onClick={this.handleUndo.bind(this)} disabled={!this.props.canUndo}>
          Undo
        </button>
        <button type="button" onClick={this.handleRedo.bind(this)} disabled={!this.props.canRedo}>
          Redo
        </button>
      </>
    );
  }
}

const mapStateToProps = (state) => ({
  board: state.board.present.board,
  posDiskWhite: state.disks.present.posDisksWhite,
  posDiskBlack: state.disks.present.posDisksBlack,
  currentPlayer: state.players.currentPlayer,
  // Redux uno/redo
  canUndo: state.board.past.length > 1,
  canRedo: state.board.future.length > 0,
});

const mapDispatchToProps = {
  setBoard,
  setAllowedCells,
  addDisksPlayers,
  changeTurn,
  onUndo: () => UndoActionCreators.undo(),
  onRedo: () => UndoActionCreators.redo(),
};

export default connect(mapStateToProps, mapDispatchToProps)(Board);
