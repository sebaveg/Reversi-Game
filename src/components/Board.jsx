import React from 'react';

import Cell from '../containers/Cell';

import '../assets/styles/Board.css';

const Board = ({ board = [] }) => (
  <table className="Board">
    <tbody>
      {board.map((row, x) => (
        <tr key={x}>
          {board[x].map((cell, y) => (
            <Cell key={`${x}${y}`} disk={cell.disk} allowed={cell.allowedCell} position={[x, y]}>
              {cell.id}
            </Cell>
          ))}
        </tr>
      ))}
    </tbody>
  </table>
);
export default Board;
