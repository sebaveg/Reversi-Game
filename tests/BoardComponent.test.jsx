/* eslint-disable no-undef */
import React from 'react';
import Board from '../src/components/Board';

// let wrapper;
// beforeEach(() => {
//   wrapper = shallow(<GameScreen />);
// });

const wrapper = shallow(
  <Board />,
);

describe('<Board /> Component', () => {
  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('should render one <table>', () => {
    expect(wrapper.find('table')).toHaveLength(1);
  });
  it('set prop board should be render 64 cells', () => {
    const board = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H'];
    for (let i = 0; i < 8; i += 1) {
      board[i] = ['1', '2', '3', '4', '5', '6', '7', '8'].map((row) => ({
        id: row + board[i],
        disk: null, // white or black
        allowedCell: false, // says if enable or disable click for put disk
      }));
    }
    wrapper.setProps({ board });
    expect(wrapper.find('table')).toHaveLength(1);
    expect(wrapper.find('tbody')).toHaveLength(1);
    expect(wrapper.find('tr')).toHaveLength(8);
    // expect(wrapper.find('td')).toHaveLength(64);
  });
});
