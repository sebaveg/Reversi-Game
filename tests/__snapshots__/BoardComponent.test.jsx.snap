// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board /> Component renders correctly 1`] = `
<table
  className="Board"
>
  <tbody />
</table>
`;
